### PR TITLE
Remove subtyping

### PR DESCRIPTION
**Pull request by rossberg**
Per the vote on #69, this PR removes subtyping from the proposal. List of changes:

* Syntax:
  - remove `nullref` type
  - rename `anyref` type to `externref`
  - extend `ref.null` and `ref.is_null` instructions with new immediate of the form `func` or `extern` (this will later have to generalise to a `constype` per the [typed references proposal](https://github.com/WebAssembly/function-references))
* Typing rules:
  - `ref.null`, `ref.is_null`: determine reference type based on new immediate
  - `select`, `call_indirect`, `table.copy`, `table.init`: drop subtyping
  - `br_table`: revert to rule requiring same label types
  - `elem` segment: drop subtyping
  - `global` import: drop subtyping (link time)
* Remove subtyping rules and bottom type.
* Revert typing algorithm (interpreter and spec appendix).
* JS API:
  - remove `"nullref"`
  - rename `"anyref"` to `"externref"`
* Scripts:
  - rename `ref` result to `ref.extern`
  - rename `ref.host` value to `ref.extern`
  - drop subtyping from invocation type check
* JS translation:
  - extend harness with separate eq functions for each ref type
* Adjust tests:
  - apply syntax changes
  - remove tests for subtyping
  - change tests exercising subtyping in other ways

### CODE REVIEW

File: test/core/ref_func.wast [hunk 1]
```diff
@@ -9,8 +9,8 @@
     (i32.add (local.get $x) (i32.const 1))
   )
 
-  (global anyref (ref.func $f))
-  (global anyref (ref.func $g))
+  (global externref (ref.null extern))
```
Comment (binji) [inline test/core/ref_func.wast:12]: can be removed now?

File: document/js-api/index.bs [hunk 2]
```diff
@@ -902,9 +898,7 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
     1. If |valuetype| equals [=𝗂𝟨𝟦=], return [=𝗂𝟨𝟦.𝖼𝗈𝗇𝗌𝗍=] 0.
     1. If |valuetype| equals [=𝖿𝟥𝟤=], return [=𝖿𝟥𝟤.𝖼𝗈𝗇𝗌𝗍=] 0.
     1. If |valuetype| equals [=𝖿𝟨𝟦=], return [=𝖿𝟨𝟦.𝖼𝗈𝗇𝗌𝗍=] 0.
-    1. If |valuetype| equals [=nullref=], return [=ref.null=].
     1. If |valuetype| equals [=anyref=], return [=ToWebAssemblyValue=](undefined, |valuetype|).
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/js-api/index.bs]: Done.

File: interpreter/README.md [hunk 3]
```diff
@@ -178,13 +178,14 @@ var: <nat> | <name>
...
 relop: eq | ne | lt | ...
-sign:  s|u
+sign:  s | u
 offset: offset=<nat>
 align: align=(1|2|4|8|...)
 cvtop: trunc | extend | wrap | ...
 
 num_type: i32 | i64 | f32 | f64
-ref_type: anyref | funcref | nullref
+ref_kind: any | func
```
[the remark this answers is not in the document]
Comment (rossberg) [reply interpreter/README.md]: Done.

File: proposals/reference-types/Overview.md [hunk 4]
```diff
@@ -47,8 +46,8 @@ Notes:
 
 Typing extensions:
 
-* Introduce `anyref`, `funcref`, and `nullref` as a new class of *reference types*.
-  - `reftype ::= anyref | funcref | nullref`
+* Introduce `funref` and `externref` as a new class of *reference types*.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply proposals/reference-types/Overview.md]: Done.

File: test/core/linking.wast [hunk 5]
```diff
@@ -94,60 +94,33 @@
...
 (register "Mref_ex" $Mref_ex)
 
 (module $Mref_im
-  (global (import "Mref_ex" "g-const-null") nullref)
-  (global (import "Mref_ex" "g-const-null") funcref)
-  (global (import "Mref_ex" "g-const-null") anyref)
   (global (import "Mref_ex" "g-const-func") funcref)
-  (global (import "Mref_ex" "g-const-func") anyref)
-  (global (import "Mref_ex" "g-const-any") anyref)
+  (global (import "Mref_ex" "g-const-any") externref)
```
[the remark this answers is not in the document]
Comment (rossberg) [reply test/core/linking.wast]: Done.

File: test/core/ref_func.wast (hunk 1 above)
Comment (rossberg) [reply test/core/ref_func.wast]: Done.

**Pull request merged**